### ISSUE TITLE
[Question] Request 404 per Controller per Method

### ISSUE THREAD

**Issue by alberk8** [prometheus-net/prometheus-net#297]
Hi,
I have tried but failed as to how to enable 404  and other error code for controller and Action calls.  What I get is an overall 404.

Any help would be appreciated.

`# TYPE http_request_duration_seconds histogram
http_request_duration_seconds_sum{code="404",method="GET",controller="",action="" } 0.0002027
http_request_duration_seconds_count{code="404",method="GET",controller="",action="" } 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.001"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.002"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.004"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.008"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.016"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.032"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.064"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.128"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.256"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.512"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="1.024"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="2.048"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="4.096"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="8.192"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="16.384"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="32.768"} 1
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="+Inf"} 1`

**Comment by GlennSillsUKG** [comment]
This might be impossible. After all the page cannot be found so the request will never be forwarded to a controller.

**Comment by yazdan** [comment]
Same issue here and I'm not sure what is wrong
My sample metrics are shown below. As you can see `request_total` has correct values and what starts with `/api/` is redirected to a controller. But the data in `http_request_duration_seconds_count` and `http_requests_received_total` are all 404
```
http_request_duration_seconds_sum{code="404",method="GET",controller="",action=""} 0.027709900000000003
http_request_duration_seconds_count{code="404",method="GET",controller="",action=""} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.001"} 13
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.002"} 13
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.004"} 14
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.008"} 14
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.016"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.032"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.064"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.128"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.256"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="0.512"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="1.024"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="2.048"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="4.096"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="8.192"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="16.384"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="32.768"} 16
http_request_duration_seconds_bucket{code="404",method="GET",controller="",action="",le="+Inf"} 16
```
```
http_requests_received_total{code="404",method="GET",controller="",action=""} 16
```
and
```
request_total{path="/f/contents/7-1-1-fact_list_2__iconurl-398311.png",method="GET",status="404"} 4
request_total{path="/api/FAQController/GetAll",method="POST",status="200"} 2
request_total{path="/api/General/RatesAndReviews",method="GET",status="200"} 2
request_total{path="/f/contents/7-1-1-fact_list_0__iconurl-589481.png",method="GET",status="404"} 4
request_total{path="/f/contents/7-1-1-fact_list_1__iconurl-206508.png",method="GET",status="404"} 4
request_total{path="/api/Doctors/Journeys",method="POST",status="200"} 2
request_total{path="/f/contents/7-1-1-introduction_imageurl-486017_t_0_75_o.jpeg",method="GET",status="404"} 4
request_total{path="/api/Content/GetContentsByModel",method="POST",status="200"} 2
```

**Comment by GlennSillsUKG** [comment]
@yazdan Yeah, what you have makes sense  to me. In my case I was looking at some requests where the path was not found like /api/SomePathThatDoesNotExist. So, of course, the controller was set to "" in that case because there is no SomePathThatDoesNotExistController mapped. In my case this was happening because we had some software that was pinging an URL that did not exist.  On the other hand, your request like 

`"/f/contents/7-1-1-introduction_imageurl-486017_t_0_75_o.jpeg`  failed.

I suspect that you have a static content handler wired up and _it_ is checking the path and returning a 404 when it is unable to find the file in question. It would never map to a controller, since it isn't a Web API per se.

I didn't write this package though so you might want to take what I say with a grain of salt.

**Comment by yazdan** [comment]
@GlennSillsUKG  Thanks for your response.

I was the root cause. There was a SSO middleware that was added before `UseHttpMetrics` called.

For 404's you are correct.
For having static content served by api server. It was because I was running it on my machine it was not configured correctly. 

Again thanks

**Comment by sandersaares** [comment]
There is now `options.AddCustomLabel()` which may help cover gaps in cases where you want custom labels to differentiate metrics that do not have controllers/actions. There is also support for automatically adding `endpoint` and `page` labels. This is available in v6.0.0 prerelease package at the moment. Any feedback is most welcome, especially during preview!

As it sounds like the comments above have accurately described the situation, I will close this as "by design and/or improved support exists in newer version".
